src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs

```diff
@@ -591,64 +709,55 @@ private static async ValueTask WriteGatherAtOffsetMultipleSyscallsAsync(SafeFile
             }
         }
 
-        private static ValueTask WriteGatherAtOffsetSingleSyscallAsync(SafeFileHandle handle, IReadOnlyList<ReadOnlyMemory<byte>> buffers, long fileOffset, int totalBytes, CancellationToken cancellationToken)
+        private static async ValueTask WriteGatherAtOffsetSingleSyscallAsync(SafeFileHandle handle, MemoryHandle[] pinnedBuffers, long fileOffset, int totalBytes, CancellationToken cancellationToken)
         {
-            if (buffers.Count == 1)
-            {
-                return WriteAtOffsetAsync(handle, buffers[0], fileOffset, cancellationToken);
-            }
+            // "The array must contain enough elements to store nNumberOfBytesToWrite bytes of data, and one element for the terminating NULL. "
+            int buffersCount = pinnedBuffers.Length;
+            long[] fileSegments = new long[buffersCount + 1];
+            fileSegments[buffersCount] = 0;
 
-            return Core(handle, buffers, fileOffset, totalBytes, cancellationToken);
+            GCHandle pinnedSegments = default;
```

Comment:
I replaced the `MemoryHandle` with the lighter `GCHandle` (we always pin an array), and the methods that perform the actual syscalls not take an `IntPtr` to the array's `AddrOfPinnedObject`.

Reply:
> I replaced the MemoryHandle with the lighter GCHandle (we always pin an array)

I've not looked at the PR yet, but just responding to this comment..

That's not necessarily lighterweight.  If the Memory provided was, for example, created with MemoryMarshal.CreateFromPinnedArray, Memory.Pin will be a nop but GCHandle will still have measurable cost.

If this Memory is coming from user code, please revert this change.

Reply:
We created the array a couple of lines earlier; it's a temporary one.

The user-provided memories are still pinned with the help of `MemoryHandles`.

Reply:
Allocate the memory using `NativeMemory.Alloc`? You do not need to worry about any pinning if you do that.

Reply:
That's even better, thanks, it made the code simpler as well.